Stream Layer v32 (#7)
streamr-dev/streamr-specs: PROTOCOL.md

`errorMessage` | `string` | Human-readable message describing the error.
`errorCode` | `string` | Machine-readable string describing the type of error.

## Stream Layer

This document describes Stream Layer version `32`.

Stream Layer messages are the actual content in streams. They are identified by a [`MessageID`](#messageid), are chained together via a reference to the previous message, [`MessageRef`](#messageref) to establish message order, they are cryptographically signed by their publisher. The most prominent Stream Layer message is the [`StreamMessage`](#streammessage), which is used to transport arbitrary application-level messages from publishers to subscribers. Additionally, there are a few other message types used for requesting and delivering encryption keys.

It's worth noting that the following Control Layer messages contain a Stream Layer message: `PublishRequest`, `BroadcastMessage`, and `UnicastMessage`.

All Stream Layer messages have the following structure:

 ```

>  [version, msgId, prevMsgRef, messageType, contentType, encryptionType, encryptionKeyId, content, signatureType, signature]

Reviewer: ```suggestion
 [version, msgId, prevMsgRef, messageType, contentType, encryptionType, groupKeyId, content, signatureType, signature]
```

IMO we **should** rename all the `groupKey` stuff to `encryptionKey` since  "groupKey" being related to encryption at all is not immediately obvious, unlike "encryptionKey", at least to me.

Reviewer: (also needs same change in PR description)

Author: Yeah this crossed my mind as well. I started making the change, but then aborted. I didn't want to snowball to renaming the `GroupKeyRequest` message types as well.

I also thought that the protocol probably should not be implemented by someone who doesn't know the basics of cryptography. So keeping the `GroupKeyRequest` seems fine from that point of view.

But actually now, with the latest commit I added that the `groupKeyId` field could in case of RSA encryption be set to the recipient's public key. That conflicts with the terminology, since a group key means a shared symmetric key, so in case the field can be used to point to a RSA key pair as well, it would not be a good name for the field.

I'll have another pass and see what might work best!

Author: I don't want to snowball this at this time, so leaving the naming as is for now. "Group key" isn't by any means incorrect, it's even "more" correct (accurate) than "encryption key". Can be reconsidered in the future.

The important thing IMO here is that we make sure that the interface of the libraries are friendly enough to be used by users not familiar with cryptography. The spec doesn't need to be written in layman terms.

The only thing that's "wrong" now is that we're "abusing" the `groupKeyId` field to identify the RSA key used in case of RSA encryption. That is not a group key. But even so, I think I'll leave it as is for now.

 `encryptionType`| `number` | Encryption type as defined by the table below.
 `groupKeyId`| `string` | Identifies the symmetric key used by the publisher to encrypt the message content. `null` if the message is unencrypted.
 `content` | `string` | Content data of the message. Depends on the `messageType` how the content should be handled.
 `signatureType` | `number` | Signature type as defined by the table below.
 `signature` | `string` | Signature of the message, signed by the producer. Encoding depends on the signature type.

The various type fields have the following possible values:

#### `messageType`

`messageType` | Description
-------------- | --------
27 | [StreamMessage](#streammessage)
28 | [GroupKeyRequest](#groupkeyrequest)
29 | [GroupKeyResponse](#groupkeyresponse)
30 | [GroupKeyReset](#groupkeyreset)
31 | [GroupKeyErrorResponse](#groupkeyerrorresponse)
32 | [GroupKeyRotate](#groupkeyrotate)

#### `contentType`

Describes how the content (post-decryption in case it's encrypted) should be parsed.

`contentType` | Description
------------- | -----------

> 0 | JSON

Reviewer: Great addition. 

Maybe should define these straight up using the same format as HTTP? i.e. use mime type + also define the charset used so as to avoid any ambiguity https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type

Reviewer: Also wondering if we should have an escape hatch in addition to a hard-coded list?

Author: > use mime type + also define the charset used

Need to keep the messages as compact as possible, and having predefined numeric values seems more efficient than text strings like `application/json`. 

Charset only comes into play when converting to/from binary, which actually isn't happening here for `StreamMessage` content separately (binary conversion happens for the whole message).

This file currently defines only the messages but not their binary serialization. I added that.

Later if we switch to a binary protocol, then the `content` would of course be serialized on its own, meaning the charset needs to become explicit for the `contentType`. Perhaps then `JSON` will be redefined to `JSON_UTF8` or similar.

Author: > Also wondering if we should have an escape hatch in addition to a hard-coded list?

I didn't understand this one, can you elaborate what you mean?

Reviewer: > Charset only comes into play when converting to/from binary

Charset is always in play? JSON is **supposed** to always be in UTF{8,16,32} but if an intermediary accidentally re-encodes using ISO-8859-1 or some other charset, then there may be issues decoding the text correctly.

On this note, we should probably have a test in each client & the testing-tool that confirms valid strings like "κόσμε" are decoded correctly by the clients (see https://www.w3.org/2001/06/utf-8-wrong/UTF-8-test.html and https://www.w3.org/2001/06/utf-8-test/UTF-8-demo.html), and don't get corrupted going in and out of storage.

Reviewer: > having predefined numeric values seems more efficient than text strings like application/json.

Agreed, not suggesting including the entire mimetype in the protocol (but perhaps as an optional escape hatch), just suggesting we define the numeric types using mimetype + char encoding instead of just "JSON".

Reviewer: Added an character encoding integration test to the JS client here, and it seems to work. 👍  https://github.com/streamr-dev/streamr-client-javascript/pull/151

Author: What I meant was:

- `contentType` describes how the (text in!) `content` should be interpreted

- `content` is encoded together with the whole surrounding message (with a charset defined for the whole protocol, which is utf8)

What follows from the above is that it's not useful (or possible) for `content` to have a separate encoding.

So this is a limitation that stems from the nature of being a text-based protocol.

Sure, one could send non-utf8 content by sending arbitrary bytes in the content by encoding binary to text, but that would be wildly inefficient. And that's what we're already doing for encrypted content and signatures (and yeah it's inefficient).

So to enable full flexibility and get rid of the inefficiencies of a text protocol, in the medium term we definitely need to move to a binary-first protocol, probably protobuf-based. But first things first!

Reviewer: > Sure, one could send non-utf8 content by sending arbitrary bytes in the content by encoding binary to text, but that would be wildly inefficient

I'm talking about about special/international charset support for text, not about encoding different data types. i.e. text can use utf8, uc2, utf16, ISO-8859-1 (latin1), Windows-1252, ascii, ansi, etc. Some tools (e.g. mysql) will default to latin1 (the default charset for html) or ascii if not explicitly told to use utf8.

I'm just recommending that the protocol be totally explicit about which charset should be used so nobody can complain we didn't specify a charset if/when their non-utf8 text gets mangled by an implicit conversion to utf8.

Reviewer: Ahh I see now: https://github.com/streamr-dev/streamr-specs/pull/7/commits/1d8fe8b59b4bbb674b860ee9b9e5cf06e54c64b6

> The JSON messages are UTF8 encoded when transported under an underlying binary protocol (which from the perspective of the message definitions is undefined, but in practice the network uses websocket and in the future WebRTC).  

Perfect 👍

Other content types, including binary types, will be defined in the future.

#### `encryptionType`

> For each non-zero `encryptionType`, the `contentType` is a hex-encoded string.

Reviewer: I think this is saying the implicit `contentType` pre-decryption is a hex-encoded string, which means that the content itself will be a hex-encoded string, but I had to read this a few times to get that.

Author: Oops, should say `content` instead of `contentType`. I'll also clarify this sentence.

> For each non-zero `encryptionType`, the `contentType` is a hex-encoded string.

Reviewer: The `contentType` perhaps shouldn't be mentioned at all e.g. something like:

```suggestion
For each non-zero `encryptionType`, the `content` will be a hex-encoded binary string, encrypted using the specified `groupKeyId`.
```

-------------- | ---- |------------ | -----------------------
0 | `NONE` | No signature. signature field is empty in this case. | None.
1 | `ETH_LEGACY` | Ethereum signature produced by old clients (Message Layer version 29). The signature field is encoded as a hex string. | `streamId`, `streamPartition`, `timestamp`, `publisherId`, `content`
2 | `ETH` | Ethereum signature produced by current clients (since Message Layer version 30). The signature field is encoded as a hex string. | all the `msgId` fields, (`streamId`, `streamPartition`, `timestamp`, `sequenceNumber`, `publisherId`, `msgChainId`), `prevMsgRef`, `content`
Describes how the content (post-decryption in case it's encrypted) should be parsed.

`contentType` | Description
------------- | -----------
0 | JSON

Other content types, including binary types, will be defined in the future.

#### `encryptionType`

For each non-zero `encryptionType`, the `contentType` is a hex-encoded string.

`encryptionType` | Description
-------------- | --------
0 | Unencrypted, plaintext message.
RSA | Content is asymetrically encrypted using RSA.
AES | Content is symmetrically encrypted using AES.

> NEW_KEY_AND_AES | The content contains a concatenation of a new AES key and the content, encrypted with the old key.

Reviewer: 👍

> [32, [...msgIdFields], [...msgRefFields], 27, 0, 0, "{\"content\":42}", 2, "0x29c057786Fa..."]

Reviewer: The JSON message doesn't need to be inside the `"content":` key, right? i.e. example of "content" shouldn't include a JSON object just containing a "content" key, lest people think this is part of the protocol. It isn't, is it?

Author: Clarified

```
[32, [...msgIdFields], [...msgRefFields], 29, 0, 0, ["requestId", "streamId", ["keyId","123abc"]], 2, "0x29c057786Fa..."]
```

### GroupKeyReset

Sent by publishers to each subscribers' key exchange streams whenever they want to re-key the stream (in order to revoke access from some subscribers). 

The requestor should handle this message by adding the contained keys to their key storage.

`GroupKeyReset`s are encrypted with (`encryptionType 1 (RSA)`) and are sent to subscribers whose `rsaPublicKey` is known by the publisher (due to receiving a `GroupKeyRequest` earlier, for example. The `content` encoded as `contentType 0` (JSON) is as follows:

```
[streamId, groupKeyId, groupKey]
```

 Field        | Type      | Description
 ------------ | --------- | -----------
 `streamId`   | `string`  | The stream for which a key is being delivered.
 `groupKeyId` | `string`  | The id of the `groupKey`.
 `groupKey`   | `string`  | The group key as a hex-encoded binary string.
 
Example of a `GroupKeyReset` including the rest of the `Stream Layer` fields:

```
[32, [...msgIdFields], [...msgRefFields], 30, 0, 0, ["keyId", "123abc"], 2, "0x29c057786Fa..."]
```

### GroupKeyErrorResponse

> In case there's an error processing a `GroupKeyRequest`, this message gets sent by a publisher to the requestor's key exchange stream. 

Reviewer: Is this what would be sent if the `groupKeyId` doesn't exist or the user doesn't have authorisation for that key e.g. user product subscription time expired?

Author: Yes, exactly. Clarified.

`GroupKeyErrorResponse`s are always unencrypted (`encryptionType 0`). The `content` encoded as `contentType 0` (JSON) is as follows:

```
[requestId, streamId, errorCode, errorMessage]
```

 Field        | Type      | Description
 ------------ | --------- | -----------
 `requestId`  | `string`  | The `requestId` of the `GroupKeyRequest`, identifying which request is being responded to.

>  `streamId`   | `string`  | The stream for which a key is being delivered.

Reviewer: perhaps should include the `groupKeyId` in the error, clients can figure this out by looking at which request corresponds to which groupKeyId, but might help debugging.

Author: The request can have multiple `groupKeyIds`. It would make sense to include in the error response the ones that errored. I'll make that change.

This actually leads us to a situation where _some_ of the requested ids may succeed and some may fail. Perhaps in this case the publisher should send both a `GroupKeyResponse` and `GroupKeyErrorResponse`. I guess that's fine?!

Author: Added some text about that, and also considered the case where the publisher has many instances connected to the network and subscribed to the key exchange stream. In this case the requestor will receive many responses for the same request.

 `errorCode`  | `string`  | An error code to help the subscriber categorize the error
 `errorMessage`| `string` | An error message intended for a human, explaining what went wrong
 
Example of a `GroupKeyErrorResponse` including the rest of the `Stream Layer` fields:

```
[32, [...msgIdFields], [...msgRefFields], 31, 0, 0, ["requestId", "streamId", "ERROR_CODE", "Error message"], 2, "0x29c057786Fa..."]
```

### GroupKeyRotate

This message is sent by publishers to rotate the encryption key and deliver it to existing subscribers (for forward secrecy). Subscribers should add the contained key into their key storage. The message is encrypted with another (previous) key. Publishers usually send this message to inform subscribers of a new key in advance, so that they every subscriber doesn't need to send a `GroupKeyRequest` to the publisher when they encounter the first `StreamMessage` encrypted with the new key.

> `GroupKeyRotate`s are always encrypted (`encryptionType 2 (AES)`) with a previous key. The `content` encoded as `contentType 0` (JSON) is as follows:

Reviewer: with "a previous key" i.e. any previous key, or "the previous key" i.e. the key used to encrypt the immediately proceeding message? Does it matter? Maybe should be explicit if you mean "any"

Author: I mean the immediately preceding key. Clarified

### MessageRef

> Used inside a Stream Layer message to identify the previous message on the same `msgChainId` (defined above). If the `MessageRef` is not provided by the publisher, then the stream will have weaker deliver guarantees, as subscribers will not be able to detect missing messages and request gapfills.

Reviewer: ```suggestion
Used inside a Stream Layer message to identify the previous message on the same `msgChainId` (defined above). If the `MessageRef` is not provided by the publisher, then the stream will have weaker delivery guarantees, as subscribers will not be able to detect missing messages and request gapfills.
```
